File: backend/github_fetcher.py
"""
github_fetcher.py

Fetches source code files from a public GitHub repository
and returns them combined into a single string for a one-shot Gemini review.
"""

import os
import re
import time
import requests
from concurrent.futures import ThreadPoolExecutor, as_completed
from dotenv import load_dotenv

load_dotenv()

GITHUB_TOKEN = os.getenv("GITHUB_TOKEN")

HEADERS = {"Accept": "application/vnd.github+json"}
if GITHUB_TOKEN:
    HEADERS["Authorization"] = f"Bearer {GITHUB_TOKEN}"

EXT_LANGUAGE_MAP = {
    "py": "python",
    "js": "javascript", "jsx": "javascript",
    "ts": "typescript", "tsx": "typescript",
    "cpp": "cpp", "cc": "cpp", "cxx": "cpp",
    "c": "c", "h": "c",
    "cs": "csharp",
    "java": "java",
    "go": "go",
    "rs": "rust",
    "kt": "kotlin",
    "php": "php",
    "rb": "ruby",
    "swift": "swift",
    "sh": "shell", "bash": "shell",
    "html": "html", "htm": "html",
    "css": "css", "scss": "scss",
    "sql": "sql",
    "json": "json",
    "yaml": "yaml", "yml": "yaml",
    "xml": "xml",
    "md": "markdown",
}

SKIP_PATH_PATTERNS = [
    "node_modules/", ".git/", "dist/", "build/", "__pycache__/",
    ".venv/", "venv/", ".mypy_cache/", "coverage/",
    "vendor/", "target/", "bin/", "obj/",
    "package-lock.json", "yarn.lock", "poetry.lock",
]

MAX_FILE_SIZE_BYTES = 60 * 1024    # 60 KB per file
MAX_FILES = 20                      # Max files fetched
MAX_COMBINED_CHARS = 30_000        # Hard cap before sending to Gemini
MAX_WORKERS = 5
REQUEST_TIMEOUT = 12


# --- Exceptions ---

class GitHubFetchError(Exception):
    pass

class RepoNotFoundError(GitHubFetchError):
    pass

class RateLimitError(GitHubFetchError):
    pass


# --- Helpers ---

def parse_github_url(repo_url: str) -> tuple[str, str, str | None]:
    url = repo_url.strip().rstrip("/")
    match = re.match(r"https?://github\.com/([^/]+)/([^/]+)/tree/([^/]+)/?$", url)
    if match:
        return match.group(1), match.group(2).removesuffix(".git"), match.group(3)
    match = re.match(r"https?://github\.com/([^/]+)/([^/]+?)(?:\.git)?$", url)
    if match:
        return match.group(1), match.group(2), None
    raise GitHubFetchError(f"Invalid GitHub URL: '{repo_url}'")


def _get(url: str, retries: int = 3) -> requests.Response:
    delay = 2
    for attempt in range(retries):
        try:
            resp = requests.get(url, headers=HEADERS, timeout=REQUEST_TIMEOUT)
        except requests.RequestException as e:
            if attempt == retries - 1:
                raise GitHubFetchError(f"Network error: {e}") from e
            time.sleep(delay); delay *= 2; continue

        if resp.status_code == 200:
            return resp
        if resp.status_code == 404:
            raise RepoNotFoundError(f"Repository not found: {url}")
        if resp.status_code in (403, 429):
            reset = resp.headers.get("X-RateLimit-Reset", "unknown")
            raise RateLimitError(
                f"GitHub rate limit exceeded. Reset at: {reset}. "
                "Add GITHUB_TOKEN to .env for higher limits."
            )
        if resp.status_code >= 500 and attempt < retries - 1:
            time.sleep(delay); delay *= 2; continue
        resp.raise_for_status()

    raise GitHubFetchError(f"Failed to GET {url} after {retries} attempts.")


def _resolve_default_branch(owner: str, repo: str) -> str:
    url = f"https://api.github.com/repos/{owner}/{repo}"
    return _get(url).json().get("default_branch", "main")


def _get_tree(owner: str, repo: str, branch: str) -> list[dict]:
    url = f"https://api.github.com/repos/{owner}/{repo}/git/trees/{branch}?recursive=1"
    data = _get(url).json()
    if data.get("truncated"):
        print(f"WARNING: Repo tree truncated at {len(data['tree'])} items.")
    return data.get("tree", [])


def _ext_from_path(path: str) -> str:
    return path.rsplit(".", 1)[-1].lower() if "." in path else ""


def _should_skip(path: str) -> bool:
    return any(p in path for p in SKIP_PATH_PATTERNS)


def _fetch_raw(owner: str, repo: str, branch: str, path: str) -> str | None:
    url = f"https://raw.githubusercontent.com/{owner}/{repo}/{branch}/{path}"
    try:
        return _get(url).text
    except Exception as e:
        print(f"  Skipping '{path}': {e}")
        return None


# --- Main functions ---

def fetch_repo_files(repo_url: str) -> list[dict]:
    """
    Fetch individual file dicts. Kept for advanced use.
    Returns: [{"path", "content", "language", "size"}, ...]
    """
    owner, repo, branch = parse_github_url(repo_url)
    if not branch:
        branch = _resolve_default_branch(owner, repo)

    tree = _get_tree(owner, repo, branch)

    candidates = sorted([
        item for item in tree
        if item.get("type") == "blob"
        and _ext_from_path(item["path"]) in EXT_LANGUAGE_MAP
        and item.get("size", 0) <= MAX_FILE_SIZE_BYTES
        and not _should_skip(item["path"])
    ], key=lambda x: x["path"])[:MAX_FILES]

    results = []
    with ThreadPoolExecutor(max_workers=MAX_WORKERS) as executor:
        futures = {
            executor.submit(_fetch_raw, owner, repo, branch, item["path"]): item
            for item in candidates
        }
        for future in as_completed(futures):
            item = futures[future]
            content = future.result()
            if content:
                ext = _ext_from_path(item["path"])
                results.append({
                    "path": item["path"],
                    "content": content,
                    "language": EXT_LANGUAGE_MAP.get(ext, "text"),
                    "size": len(content.encode())
                })

    results.sort(key=lambda x: x["path"])
    return results


def fetch_combined_code(repo_url: str) -> tuple[str, list[str]]:
    """
    Fetch all code files from a repo and combine into a SINGLE string
    formatted for one Gemini call. Enforces the MAX_COMBINED_CHARS cap.

    Returns:
        combined_code (str): All file contents joined with headers
        file_paths (list[str]): List of included file paths
    """
    files = fetch_repo_files(repo_url)

    if not files:
        raise GitHubFetchError("No reviewable source files found in this repository.")

    combined = ""
    included_paths = []

    for f in files:
        header = f"\n\n{'='*60}\n// FILE: {f['path']}  [{f['language']}]\n{'='*60}\n"
        block = header + f["content"]

        if len(combined) + len(block) > MAX_COMBINED_CHARS:
            print(f"  Token cap reached. Skipping remaining files (stopped at {f['path']}).")
            break

        combined += block
        included_paths.append(f["path"])

    print(f"Combined {len(included_paths)} files → {len(combined):,} chars (cap: {MAX_COMBINED_CHARS:,})")
    return combined, included_paths

File: backend/githubFetcher.js
// githubFetcher.js — Fetch source code from public GitHub repositories

/**
 * Fetches source code files from a public GitHub repository
 * and returns them combined into a single string for a one-shot Gemini review.
 */

const GITHUB_TOKEN = process.env.GITHUB_TOKEN;

const HEADERS = { Accept: "application/vnd.github+json" };
if (GITHUB_TOKEN) {
  HEADERS.Authorization = `Bearer ${GITHUB_TOKEN}`;
}

const EXT_LANGUAGE_MAP = {
  py: "python",
  js: "javascript",
  jsx: "javascript",
  ts: "typescript",
  tsx: "typescript",
  cpp: "cpp",
  cc: "cpp",
  cxx: "cpp",
  c: "c",
  h: "c",
  cs: "csharp",
  java: "java",
  go: "go",
  rs: "rust",
  kt: "kotlin",
  php: "php",
  rb: "ruby",
  swift: "swift",
  sh: "shell",
  bash: "shell",
  html: "html",
  htm: "html",
  css: "css",
  scss: "scss",
  sql: "sql",
  json: "json",
  yaml: "yaml",
  yml: "yaml",
  xml: "xml",
  md: "markdown",
};

const SKIP_PATH_PATTERNS = [
  "node_modules/",
  ".git/",
  "dist/",
  "build/",
  "__pycache__/",
  ".venv/",
  "venv/",
  ".mypy_cache/",
  "coverage/",
  "vendor/",
  "target/",
  "bin/",
  "obj/",
  "package-lock.json",
  "yarn.lock",
  "poetry.lock",
];

const MAX_FILE_SIZE_BYTES = 60 * 1024; // 60 KB per file
const MAX_FILES = 20;
const MAX_COMBINED_CHARS = 30_000;
const MAX_WORKERS = 5;
const REQUEST_TIMEOUT = 12_000; // ms

// --- Custom Error Classes ---

class GitHubFetchError extends Error {
  constructor(message) {
    super(message);
    this.name = "GitHubFetchError";
  }
}

class RepoNotFoundError extends GitHubFetchError {
  constructor(message) {
    super(message);
    this.name = "RepoNotFoundError";
  }
}

class RateLimitError extends GitHubFetchError {
  constructor(message) {
    super(message);
    this.name = "RateLimitError";
  }
}

// --- Helpers ---

function parseGithubUrl(repoUrl) {
  const url = repoUrl.trim().replace(/\/+$/, "");

  // Match: https://github.com/owner/repo/tree/branch
  let match = url.match(
    /^https?:\/\/github\.com\/([^/]+)\/([^/]+)\/tree\/([^/]+)\/?$/
  );
  if (match) {
    return {
      owner: match[1],
      repo: match[2].replace(/\.git$/, ""),
      branch: match[3],
    };
  }

  // Match: https://github.com/owner/repo
  match = url.match(/^https?:\/\/github\.com\/([^/]+)\/([^/]+?)(?:\.git)?$/);
  if (match) {
    return { owner: match[1], repo: match[2], branch: null };
  }

  throw new GitHubFetchError(`Invalid GitHub URL: '${repoUrl}'`);
}

async function fetchWithRetry(url, retries = 3) {
  let delay = 2000;

  for (let attempt = 0; attempt < retries; attempt++) {
    try {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), REQUEST_TIMEOUT);

      const resp = await fetch(url, {
        headers: HEADERS,
        signal: controller.signal,
      });
      clearTimeout(timeout);

      if (resp.ok) return resp;

      if (resp.status === 404) {
        throw new RepoNotFoundError(`Repository not found: ${url}`);
      }

      if (resp.status === 403 || resp.status === 429) {
        const reset = resp.headers.get("X-RateLimit-Reset") || "unknown";
        throw new RateLimitError(
          `GitHub rate limit exceeded. Reset at: ${reset}. Add GITHUB_TOKEN to .env for higher limits.`
        );
      }

      if (resp.status >= 500 && attempt < retries - 1) {
        await new Promise((r) => setTimeout(r, delay));
        delay *= 2;
        continue;
      }

      throw new GitHubFetchError(
        `GitHub API returned ${resp.status} for ${url}`
      );
    } catch (err) {
      if (
        err instanceof GitHubFetchError ||
        err instanceof RepoNotFoundError ||
        err instanceof RateLimitError
      ) {
        throw err;
      }
      if (attempt === retries - 1) {
        throw new GitHubFetchError(`Network error: ${err.message || err}`);
      }
      await new Promise((r) => setTimeout(r, delay));
      delay *= 2;
    }
  }

  throw new GitHubFetchError(`Failed to GET ${url} after ${retries} attempts.`);
}

async function resolveDefaultBranch(owner, repo) {
  const url = `https://api.github.com/repos/${owner}/${repo}`;
  const resp = await fetchWithRetry(url);
  const data = await resp.json();
  return data.default_branch || "main";
}

async function getTree(owner, repo, branch) {
  const url = `https://api.github.com/repos/${owner}/${repo}/git/trees/${branch}?recursive=1`;
  const resp = await fetchWithRetry(url);
  const data = await resp.json();
  if (data.truncated) {
    console.log(
      `WARNING: Repo tree truncated at ${data.tree?.length || 0} items.`
    );
  }
  return data.tree || [];
}

function extFromPath(path) {
  if (!path.includes(".")) return "";
  return path.split(".").pop().toLowerCase();
}

function shouldSkip(path) {
  return SKIP_PATH_PATTERNS.some((pattern) => path.includes(pattern));
}

async function fetchRaw(owner, repo, branch, path) {
  const url = `https://raw.githubusercontent.com/${owner}/${repo}/${branch}/${path}`;
  try {
    const resp = await fetchWithRetry(url);
    return await resp.text();
  } catch (err) {
    console.log(`  Skipping '${path}': ${err.message || err}`);
    return null;
  }
}

// --- Main Functions ---

/**
 * Fetch individual file dicts from a GitHub repo.
 * Returns: [{ path, content, language, size }, ...]
 */
async function fetchRepoFiles(repoUrl) {
  const { owner, repo, branch: explicitBranch } = parseGithubUrl(repoUrl);
  const branch = explicitBranch || (await resolveDefaultBranch(owner, repo));

  const tree = await getTree(owner, repo, branch);

  // Filter and sort candidates
  const candidates = tree
    .filter(
      (item) =>
        item.type === "blob" &&
        extFromPath(item.path) in EXT_LANGUAGE_MAP &&
        (item.size || 0) <= MAX_FILE_SIZE_BYTES &&
        !shouldSkip(item.path)
    )
    .sort((a, b) => a.path.localeCompare(b.path))
    .slice(0, MAX_FILES);

  // Fetch files in parallel batches (MAX_WORKERS at a time)
  const results = [];
  for (let i = 0; i < candidates.length; i += MAX_WORKERS) {
    const batch = candidates.slice(i, i + MAX_WORKERS);
    const batchResults = await Promise.all(
      batch.map(async (item) => {
        const content = await fetchRaw(owner, repo, branch, item.path);
        if (content) {
          const ext = extFromPath(item.path);
          return {
            path: item.path,
            content,
            language: EXT_LANGUAGE_MAP[ext] || "text",
            size: Buffer.byteLength(content, "utf8"),
          };
        }
        return null;
      })
    );
    results.push(...batchResults.filter(Boolean));
  }

  results.sort((a, b) => a.path.localeCompare(b.path));
  return results;
}

/**
 * Fetch all code files from a repo and combine into a SINGLE string
 * formatted for one Gemini call. Enforces the MAX_COMBINED_CHARS cap.
 *
 * @returns {{ combinedCode: string, filePaths: string[] }}
 */
async function fetchCombinedCode(repoUrl) {
  const files = await fetchRepoFiles(repoUrl);

  if (!files.length) {
    throw new GitHubFetchError(
      "No reviewable source files found in this repository."
    );
  }

  let combined = "";
  const includedPaths = [];
  let truncated = false;

  for (const f of files) {
    const header = `\n\n${"=".repeat(60)}\n// FILE: ${f.path}  [${f.language}]\n${"=".repeat(60)}\n`;
    const block = header + f.content;

    if (combined.length + block.length > MAX_COMBINED_CHARS) {
      console.log(
        `  Token cap reached. Skipping remaining files (stopped at ${f.path}).`
      );
      truncated = true;
      break;
    }

    combined += block;
    includedPaths.push(f.path);
  }

  console.log(
    `Combined ${includedPaths.length} files → ${combined.length.toLocaleString()} chars (cap: ${MAX_COMBINED_CHARS.toLocaleString()})`
  );

  return { combinedCode: combined, filePaths: includedPaths, truncated };
}

module.exports = {
  fetchCombinedCode,
  fetchRepoFiles,
  GitHubFetchError,
  RepoNotFoundError,
  RateLimitError,
};
